evm_interpreter/src/lib.rs

```diff
@@ -474,3 +475,48 @@ impl From<InternalError> for ExitCode {
         ExitCode::FatalError(e.into())
     }
 }
+
+///
+/// Charge native and a proportional amount of ergs.
+/// The ergs are computed as:
+/// native / AVERAGE_NATIVE_PER_GAS * ERGS_PER_GAS
+///
+pub fn charge_native_and_proportional_gas<R: Resources>(
+    resources: &mut R,
+    native_to_charge: u64,
+) -> Result<(), SystemError> {
+    use zk_ee::system::Computational;
+    let ergs = Ergs(
+        native_to_charge
+            .div_ceil(AVERAGE_NATIVE_PER_GAS)
+            .saturating_mul(ERGS_PER_GAS),
+    );
+    let to_charge = R::from_ergs_and_native(ergs, R::Native::from_computational(native_to_charge));
+    resources.charge(&to_charge)
+}
+
+///
+/// Helper to charge gas after the execution of a system hook.
+/// To be used when charging gas on the fly is too complex, as in
+/// the contract deployer.
+///
+pub fn post_charge_proportional_gas<R: Resources>(
+    resources: &mut R,
+    resources_before_hook: R,
+) -> Result<(), SystemError> {
+    use zk_ee::system::Computational;
+    let native_used = resources
+        .native()
+        .diff(resources_before_hook.native())
+        .as_u64();
+    let proportional_ergs_from_native = native_used
```

Comment:
Probably we have signed division here for some reason and binary verification fails because of that